model/src/main/java/org/openrewrite/gradle/marker/GradleProjectBuilder.java

```diff
@@ -119,44 +122,32 @@ private static ResolvedGroupArtifactVersion resolvedGroupArtifactVersion(Resolve
-     * Some Gradle dependency functions will have the String "unspecified" to indicate a missing value.
-     * Rewrite's dependency API represents these missing things as "null"
-     */
-    @Nullable
-    private static String unspecifiedToNull(@Nullable String maybeUnspecified) {
-        if ("unspecified".equals(maybeUnspecified)) {
-            return null;
-        }
-        return maybeUnspecified;
-    }
-
     private static Map<String, GradleDependencyConfiguration> dependencyConfigurations(ConfigurationContainer configurationContainer) {
         Map<String, GradleDependencyConfiguration> results = new HashMap<>();
         List<Configuration> configurations = new ArrayList<>(configurationContainer);
         for (Configuration conf : configurations) {
-            List<org.openrewrite.maven.tree.Dependency> requested = conf.getAllDependencies().stream()
-                    .map(dep -> dependency(dep, conf))
-                    .collect(Collectors.toList());
+            Map<GroupArtifactVersion, org.openrewrite.maven.tree.Dependency> requestedCache = new HashMap<>();
+            List<org.openrewrite.maven.tree.Dependency> requested = conf.getDependencies().stream()
+                    .map(dep -> dependency(requestedCache, dep, conf))
+                    .collect(toList());
 
             try {
-                ResolvedConfiguration resolvedConf;
+                List<org.openrewrite.maven.tree.ResolvedDependency> resolved;
                 if (conf.isCanBeResolved()) {
-                    resolvedConf = conf.getResolvedConfiguration();
+                    ResolvedConfiguration resolvedConf = conf.getResolvedConfiguration();
+
+                    Map<ResolvedGroupArtifactVersion, org.openrewrite.maven.tree.ResolvedDependency> resolvedCache = new HashMap<>();
+                    Map<GroupArtifactVersion, org.openrewrite.maven.tree.Dependency> gavToRequested = requested.stream()
+                            .collect(Collectors.toMap(GradleProjectBuilder::groupArtifactVersion, dep -> dep, (a, b) -> a));
+                    Map<GroupArtifactVersion, ResolvedDependency> gavToResolved = resolvedConf.getFirstLevelModuleDependencies().stream()
+                            .collect(Collectors.toMap(GradleProjectBuilder::groupArtifactVersion, dep -> dep, (a, b) -> a));
+                    resolved = resolved(requestedCache, resolvedCache, gavToRequested, gavToResolved);
                 } else {
-                    // Some common configuration, like "implementation" are not resolvable.
-                    // These configurations are extended from by other configurations which are resolvable
-                    // But a recipe author will be interested in where the dependency is originally coming from
-                    resolvedConf = configurationContainer.create("resolvable" + conf.getName(), it -> it.extendsFrom(conf))
-                            .getResolvedConfiguration();
+                    resolved = emptyList();
```

Comment:
Unless I'm missing something it seems like [this test](https://github.com/openrewrite/rewrite/blob/d434dae7201c3a51c3b588aa6f16aaf8ad4b3e60/rewrite-gradle/src/test/java/org/openrewrite/gradle/search/DependencyInsightTest.java#L34) would fail after these changes

If we get rid of this then a search recipe like DependencyInsight wont be able to tell what dependencies in implementation (and other unresolvable configurations) resolved to.

Reply:
So yes, you are correct. This would impact that test. 

I'm thinking that `gradle.search.DependencyInsight` needs some rework before we actually merge this thing, but I think it'll bring things more inline with the way that Gradle behaves with respect to it's same task.

When you ask Gradle to perfrom `dependencyInsight --configuration implementation`, you would actually see no results for the test in question as it's not within the bucket of dependencies as a first order element. However, if you were to run `dependencyInsight --configuration compileClasspath`, you would see that the dependency is a child of the one selected by the test.

Now with that said, I definitely see the point of wanting/needing to point back to all originating dependencies from which a queried one exists within the LST. So what I propose is something akin to the following:

1. Find the dependency within a provided configuration, exactly by default (mimic Gradle semantics)
2. Add maybe a `transitive` flag with which we'd use any resolvable configuration that extends from the specified one OR all resolvable configurations and place a search marker anywhere that dependency comes in from. This would involve being able to build a path from a first-order dependency to the one that is being searched for, then placing a marker on all first-order dependencies based upon the provided configuration OR all configurations if none was provided.